synapse/handlers/room_member.py

```diff
@@ -548,13 +549,17 @@ async def update_membership_locked(
         )
 
         if (
-            not self.allow_per_room_profiles and not is_requester_server_notices_user
-        ) or requester.shadow_banned:
+            (not self.allow_per_room_profiles and not is_requester_server_notices_user)
+            or requester.shadow_banned
+            or len(content.get("displayname", "")) > MAX_DISPLAYNAME_LEN
```

Comment:
We should probably check the avatar URL while we're here.

I also don't think this check is correct -- if the display name is too long, this shouldn't mean that we also strip off the avatar URL! (And vice versa.)

@anoadragon453 asserted in #9236 that we should "fail the request". This is silently stripping it, which I'm unsure is the right thing to do.

Reply:
Have added the avatar URL length check

Was unsure if it was safe to "fail the request" - would that essentially mean having synapse prevent the event from ever having happened? What would happen if other servers did accept it?